### ISSUE TITLE
Scientific Workflow Provenance

### ISSUE THREAD

**Issue by vsoch** [schemaorg/schemaorg#2089]
hey @schemaorg!

Thanks for your discussion so far in #2059, we are making progress and I hope the meeting with @rvguha in January can solidify some of the work and review discussion with OCI and the toy examples I've created. This is a related issue that I'm working on in parallel - and it hits many more users (including those that don't use containers). The question at hand is:

> How do we use schema.org to represent a scientific workflow?

In the context of containers, if we do have a representation of A Container in schema.org, we would simply define the relationships between them using some kind of Workflow model. Outside of containers, you can have other kinds of SoftwareApplication working together.

## Does schema.org have a model?
From my basic search here, I can't find something that cleanly exists what I'm thinking of. You could say different interactions are kinds of [Actions](https://schema.org/Action) or that a workflow is (sort of) [SoftwareSourceCode](https://schema.org/SoftwareSourceCode) with some kind of Accessibility defined... but that really isn't right. I don't see a clear "scientific workflow" representation that would be desired, to go along with Dataset. Is there initiative working on this?

## Does bioschemas have a model?
I next went back to bioschemas, and I thought I'd find simple concepts for workflows and experiments here. As far as I can tell, bioschemas is working on [biological entties](http://bioschemas.org/specifications/drafts) (akin to genes, proteins, etc.) and less so something like an experiment, a workflow, a step, an input, etc.

## What about ontology?
From @satra I was pointed to what the neuroimaging data model (nidm) community is embracing to describe neuroimaging worklows - [ProvOne](http://vcvcomputing.com/provone/provone.html) is a nicely defined ontology. Has there been discussion about representation of these models into schema.org?

## My Use Case
The Encode Data Portal (Encode-DCC) I beiieve has on the order of hundreds of thousands --> (million?) datasets with comlicated (many non container based) workflows that need to be described. Right now they have a fairly non-standarized description and need something better. In that I'm aware of the sheer number of labs / institutions with this need, and use of schema.org (again) to drive search, and the need to link Datasets with workflows, how can we best go about this?

 - I'm aware there are a million and a half actual workflow managers (e.g., cwl, wdl, or just look at [the awesomepipelines](https://github.com/pditommaso/awesome-pipeline) so we would want to represent a level of abstraction that isn't wed to a particular one, but would generally let any given pipeline export / describe using the standard.

Here are some links to share from master @satra!

 - [Nidm Workflows](https://docs.google.com/presentation/d/10WEaWARvpT05rWFEhOkxq9WCCR1jhFrwuaWQ_ZoYn0k/edit#slide=id.g1fc8e85420_1_0) is a good example of how a specific domain can use ProvOne to describe a niche datatype (brains! :jack_o_lantern: ) So for example, if schema.org can represent ProvOne, that also means there is some kind of mapping to this entire community. Note the slides on use cases and goals - they are very good.
 - [ProvOne Python](https://github.com/incf-nidash/PyNIDM/blob/master/nidm/core/provone.py) not to say that Python is the only rule of the world, but this is to show substantial work has been already done by the nidm team with respect to ProvOne. Python (imho) is the strongest language for scientific programming, as opposed to something like GoLang.

I'll leave it at that to start our discussion! Thanks everyone.

**Comment by vsoch** [comment]
And quickly, some more background on the Encode-DCC use case:

These are the (non-standardized, or N=1 it's their own standard) definitions:

 - https://www.encodeproject.org/profiles/pipeline
 - https://www.encodeproject.org/profiles/software
 - https://www.encodeproject.org/profiles/software_version 
 - https://www.encodeproject.org/profiles/analysis_step
 - https://www.encodeproject.org/profiles/analysis_step_version
 - https://www.encodeproject.org/profiles/analysis_step_run
 - https://www.encodeproject.org/profiles/file

This does a good job to show the kinds of things we want to model. A huge chunk of them (the SoftwareApplication and SoftwareVersion) I know we already have in schema.org, awesome! It's the workflow bits (step, run, pipeline) that need to somehow be tied together. 

The other interesting thing  (these are notes from Seth in my group) is that there are associated Quality Control definitions. This is obviously huge outside of just workflows, it's for any kind of Software or thing that can be tested.

 - https://www.encodeproject.org/profiles/chipseq_filter_quality_metric
 - https://www.encodeproject.org/profiles/mad_quality_metric

And also from Seth, an example page and json that is rendered from the metadata above:
 - https://www.encodeproject.org/pipelines/ENCPL272XAE/
 - https://www.encodeproject.org/experiments/ENCSR987PNT/

So while a group like ours could go and come up with some modified niche thing for our N=1 use case, we really go a lot farther enaging with others to work on (already underway coordinated efforts) that are birds of a feather. This is what I'd like to do for Encode-DCC, and I'd like this to have much broader impact than just one or two groups at Stanford, or even one domain. This is what I'd like to talk about.

**Comment by vsoch** [comment]
A little more background and things I'm learning

## Research Objects
[ResearchObjects](http://www.researchobjects.org/) is a related ontology that seems to be about components involved in publications, or mroe specifically for workflows provenance for files / folders they produce. For example, [CWL uses it](https://github.com/common-workflow-language/cwlprov) via "CWLprov" to do this. There is a [workflow description](http://wf4ever.github.io/ro/#wfdesc) that (not surprisingly!) also goes back to ProvONE I mentioned earlier.

## Workflow Focused Initiatives
 - I haven't reached out to her yet, but [Yolanda Gil](https://www.isi.edu/~gil/publications.php) seems to be quite a force in the workflow ontology space! She works on the [Disk Ontology](http://disk-project.org/ontology/disk/index-en.html#http://www.wings-workflows.org/ontology/workflow.owl#Workflow) and [OPMV-Prov](http://www.opmw.org/model/OPMW/) and [PPlan](http://www.opmw.org/model/p-plan/).
 - [EDAM](http://bioportal.bioontology.org/ontologies/EDAM/?p=classes&conceptid=http%3A%2F%2Fedamontology.org%2Fdata_0006) would do a good job to describe the data objects (inputs and outputs) and I've seen this around bioschemas.
 - [OBI](http://www.ontobee.org/ontology/catalog/OBI?iri=http://www.w3.org/2002/07/owl%23Class) looks really good for describing "biomedical stuff" - the number of detailed terms is very impressive.
 - [A really nice talk with examples](https://docs.google.com/presentation/d/e/2PACX-1vTWTJqj6zm0Gq5UeyFWUbkkjlMv0JPHAdAlHzhFnthmFKHL_VzyQi0SKKm3FLatEx8RVDrRkSezK_8I/pub?slide=id.g19045febce_1_841)

Holy cow there are a lot! Looking forward to hearing your thoughts.

**Comment by tetron** [comment]
I think it depends on what you're trying to do.  If you want something directly executable (for reproducibility) but still grounded in linked data + software containers based then you want something like Research Objects + CWL.  If you want to answer higher level queries about what kinds of things a workflow does, you want semantic markup with domain-specific ontologies like EDAM.  I don't know the value is of workflow ontologies that are not specific enough to be directly executable.

**Comment by mr-c** [comment]
The Encode use cases may also be a good fit for [BioCompute Objects](https://osf.io/h59uh/) which can interoperate with RO/CWLProv.

**Comment by vsoch** [comment]
@mr-c BioCompute looks fantastic! I'm especially happy to see a very active Github repo and community too :)  I'll run this by my group and return with any questions. Thanks for leaving the issue open as we discuss! Here are some Github links for others interested:

 - [Here is the Github Introduction](https://github.com/biocompute-objects/BCO_Specification/blob/master/introduction.md)
 - [The Specification](https://github.com/biocompute-objects/BCO_Specification/blob/master/bco-specification.md) (note all the names at the bottom!)
 - [The Repository Base](https://github.com/biocompute-objects/BCO_Specification)
